### DISCUSSION TITLE
Checks in FreeRDP

### DISCUSSION THREAD

**Discussion by hardening** [FreeRDP/FreeRDP#7019] [General]
So I'm starting a discussion after a talk with @akallabeth, I'll mainly take as example #7012 to illustrate my talk, but we had the exact same problem with RAIL "hardening" some time ago (that has been merged and would need a complete cleanup).

I think checks must be done at the public API level and not anywhere else, internal checks aren't accurate. I'm gonna argue this here.

So, first argument I've read: "compiler can remove duplicated checks and simplify things". Well I generally think that's a really bad idea to rely on the compiler to cleanup things for you. And given the number of platforms we're running on, they can be very heterogeneous. So probably the compiler, in some trivial cases, can detect these, but also sometime it's just not feasible. As an example, in #7012, the compiler can hardly remove the checks on hashtable `table` while browsing a bucket raw. So I think we should not hide behind a "the compiler will do the trick".

Second idea: "better safe than sorry". First if we take this idea and push it at its limits then we'll end up with checks all over the code even in the smallest function, so that's not a good idea. My second objection is that adding these runtime checks don't make the code safer, they make the code less readable and make error handling more vague. So a call to `HashTable_Add` will return the same `FALSE` value if: input parameters are wrong, or we're out-of-memory or if we've self detected an internal bug. And I think we don't want the same behaviour, if we've detected that we're broken, or if it's just a matter of invalid input given to us. For sure, I don't want to continue execution as if nothing happened when we've detected an unexpected behaviour internally, because that's either a corruption, or a bug on our side.
The fact that these checks confuse input checking and internal bugs leads to super nasty bugs, on the developer's side. I tend to think that a good old segfault is way better than a vague error returned, as we will catch the bug as soon at it happens in gdb (or even in a core file). With the extra unneeded checks, you will have the returned error that will pop among function calls and finally you will have a very generic error message very far from the real error position and totally out-of-context. I had such experience when playing with RAIL, it took me ages to find where was an internal bug.

IMHO we must have a policy where error returns in public API are for input parameters checks and contextual things (OOM, IO errors). And all the rest must be `assert()`, I'm totally for adding some internal sanity checks, but they should be some assertions and kinda abort the program when they fail.

So for instance:
```C
static BOOL HashTable_Equals(wHashTable* table, const wKeyValuePair* pair, const void* key)
{
       if (!key || !pair || !table)
           return FALSE;
      ....
}
```
Shall be:
```C
static BOOL HashTable_Equals(wHashTable* table, const wKeyValuePair* pair, const void* key)
{
       assert(key && pair && table);
      ....
}
```

Another point is readability: this way we know that if in the code we see some checks on the input parameters, it's because these values are kinda legitimate, the function can have these values passed as parameters, it's not a self-protection. If you want self-protecting code, then `asserts` are made for this.

And finally against `assert` I heard the fact that they expand to void when compiled in release mode, and so we will end up with segfaults in release binaries. Well, I think I've already treated that argument above, I think it's worse to have errors returned out-of-context like when you don't use assert but instead return FALSE but the code knows something's badly broken. And it's even worse to continue execution when we have detected a totally unexpected behaviour, we amplify the risk of further corruptions and bugs that would show up in funky ways.

So my proposal summarized:
* checks only on input parameters of public API;
*  `assert` in private code when we want to add some sanity checks;

**Comment by akallabeth** [comment]
I agree with the checks on input parameters of public API, but:
`assert` is missing whenever it really counts (e.g. not a debug build)

**Reply by hardening** [reply]
Can you elaborate ? On which platform ? Is it the bothering of missing debug symbols sometime ?

**Reply by akallabeth** [reply]
no, it is that assert is only defined in debug builds. e.g. see https://en.cppreference.com/w/cpp/error/assert (same is true for plain `C`)

**Reply by hardening** [reply]
Can elaborate more, what's your expectations with `assert` ? Most of the time asserts check for a non null pointer, so if they expand to void in release mode, then you end up with a segfault (but a very well located one). If you have the `assert` expanded to something, you end up with an `abort()`. I can't see a big difference between the 2, did I miss something ?

Even in release mode, does it bring us something useful to avoid a segfault by doing internal checks ? As I said in my previous post, what's the gain of continuing execution when we have detected an internal bug ?

On my side, I'm not sure I'd want all assert to be present in release build. We could define `FREERDP_ASSERT(x)` and `FREERDP_HARD_ASSERT(x)`, the first one would expand to void in release, and the second one would do always the check and would call `abort()`

**Reply by akallabeth** [reply]
your assumption with `Most of the time asserts check for a non null pointer`.
I´d want an assert that aborts (or orderly shuts down) if some violation is detected.
Nothing against some `FREERDP_ASSERT` or `WINPR_ASSERT` to achieve that, but I´d like it be enabled in release builds unless otherwise requested by some `I_KNOW_WHAT_I_DO_DO_NOT_COMPILE_ASSERTS` option

**Reply by hardening** [reply]
Ok we can agree on a build options to have assert not being void in release build.

And also sometime in winPR some asserts checks for magic values. Good point.

Anyway you've not answered my question about expectations, in practice what makes a difference in a build release between an abort and a segfault ? I'm not saying asserts with abort are never useful in release mode (they are quite useful in `glibc` when `free` or `realloc` detects corruptions), I'm just trying to understand what makes for you an `abort()` so much better compared to a segfault.

**Reply by akallabeth** [reply]
* the option that the `abort/ASSERT` call can also transport some additional information. (backtrace the log, ...)
* Segfaults are the best case (and mostly similar), but continuing with an invalid state is the other (reasonably frequent) option, so I tend to `abort` when our expectations are not met.

**Reply by akallabeth** [reply]
I´ve just encountered such a case again, where (by mistake) I´ve forgotten a `NULL` check and fixed it in #7049 
The crash log did not reveal enough data by default (of course, you can try to get a retrace, but that is often not possible).
I´d expect some check for `NULL` assumptions and a log entry if this assumption is not met.
If the app then forcefully crashes or orderly shuts down, well, that is something I do not care much about as in both cases we hit a bug and have something we can work with.

**Reply by hardening** [reply]
Well in #7049 I feel like the root problem was still some input checking: we were not checking some expectations on the remote certificate.

**Reply by akallabeth** [reply]
that is true, but it just shows what is wrong with `assert`:
1. During debugging we never encountered this case as the certificates tested against did not have `NULL` issuer or subject
2. In release builds asserts are gone, debug symbols do not exist so the stack trace is (more or less) useless
3. Having checks for expectations in code is only useful if we get the information in some fashion, so you´re right about not being verbose about which expectation was violated and where

**Reply by akallabeth** [reply]
Ok, I´ve added #7081 as an example of an assert I´d prefer (write the location and condition to `WLog` so it is part of every (proper) bug report)
